pkg/app/ops/insightcollector/collector.go

```diff
@@ -267,160 +282,162 @@ func (i *InsightCollector) getDailyInsightData(
-		},
-		{
-			Field:    "CompletedAt",
-			Operator: "<",
-			Value:    to.Unix(),
+			Value:    from,
 		},
 	}
 
-	if applicationID != "" {
-		filters = append(filters, datastore.ListFilter{
-			Field:    "ApplicationId",
-			Operator: "==",
-			Value:    applicationID,
+	var deployments []*model.Deployment
+	maxCompletedAt := to
+	for {
+		d, err := i.deploymentStore.ListDeployments(ctx, datastore.ListOptions{
+			PageSize: pageSize,
+			Filters: append(filters, datastore.ListFilter{
+				Field:    "CompletedAt",
+				Operator: "<",
+				Value:    maxCompletedAt,
+			}),
+			Orders: []datastore.Order{
+				{
+					Field:     "CompletedAt",
+					Direction: datastore.Desc,
+				},
+			},
 		})
-	}
+		if err != nil {
+			return nil, err
+		}
+		if len(d) == 0 {
+			// get all deployments in range
+			break
+		}
 
-	if projectID != "" {
-		filters = append(filters, datastore.ListFilter{
-			Field:    "ProjectId",
-			Operator: "==",
-			Value:    projectID,
-		})
+		deployments = append(deployments, d...)
+		maxCompletedAt = d[len(d)-1].CompletedAt
 	}
+	return deployments, nil
+}
 
-	deployments, err := i.deploymentStore.ListDeployments(ctx, datastore.ListOptions{
-		PageSize: pageSize,
-		Filters:  filters,
-	})
-	if err != nil {
-		i.logger.Error("failed to get deployments", zap.Error(err))
-		return &insight.ChangeFailureRate{}, time.Time{}, fmt.Errorf("failed to get deployments")
+// groupingDeployments groups deoloyments by applicationID and projectID
+func (i *InsightCollector) groupingDeployments(deployments []*model.Deployment) (map[string][]*model.Deployment, map[string][]*model.Deployment) {
+	appmap := map[string][]*model.Deployment{}
+	projectmap := map[string][]*model.Deployment{}
+	for _, d := range deployments {
+		appmap[d.ApplicationId] = append(appmap[d.ApplicationName], d)
+		projectmap[d.ProjectId] = append(projectmap[d.ApplicationName], d)
 	}
+	return appmap, projectmap
+}
 
-	if len(deployments) == 0 {
-		return &insight.ChangeFailureRate{}, time.Time{}, ErrDeploymentNotFound
+var (
+	ErrDeploymentNotFound = errors.New("deployments not found")
+)
+
+// extractDeployFrequency extracts deploy frequency from deployments with specified range
+func (i *InsightCollector) extractDeployFrequency(deployments []*model.Deployment, from, to int64, targetTimestamp int64) (*insight.DeployFrequency, []*model.Deployment) {
+	var ds []*model.Deployment
+	var rest []*model.Deployment
+	for _, d := range deployments {
+		if d.CreatedAt < to && d.CreatedAt >= from {
+			ds = append(ds, d)
+		} else {
+			rest = append(rest, d)
+		}
 	}
 
+	return &insight.DeployFrequency{
+		Timestamp:   targetTimestamp,
+		DeployCount: float32(len(ds)),
+	}, rest
+}
+
+// extractDeployFrequency extracts change failure rate from deployments with specified range
+func (i *InsightCollector) extractChangeFailureRate(deployments []*model.Deployment, from, to int64, targetTimestamp int64) (*insight.ChangeFailureRate, []*model.Deployment) {
```

Comment:
This can be an independent function, doesn't need to belong to `InsightCollector`.